internal/graphengine/graphql_go_tools_v2.go

```diff
@@ -1,3 +1,160 @@
+	if err != nil {
+		return nil, err
+	}
+
+	normalizeResult, err := parsed.Normalize()
+	if err != nil {
+		return nil, err
+	}
+
+	if !normalizeResult.Successful {
+		return nil, fmt.Errorf("error normalizing schema: %w", normalizeResult.Errors)
+	}
+
+	return parsed, nil
+}
+
+func (g graphqlGoToolsV2) handleIntrospection(schema *graphql.Schema) (res *http.Response, hijacked bool, err error) {
+	var (
+		introspectionData = struct {
+			Data introspection.Data `json:"data"`
+		}{}
+		report operationreport.Report
+	)
+	gen := introspection.NewGenerator()
+	doc, report := astparser.ParseGraphqlDocumentBytes(schema.Document())
+	if report.HasErrors() {
+		err = report
+		return
+	}
+	gen.Generate(&doc, &report, &introspectionData.Data)
+
+	var buf bytes.Buffer
+	err = json.NewEncoder(&buf).Encode(introspectionData)
+	if err != nil {
+		return
+	}
+
+	res = &http.Response{}
+	res.Body = io.NopCloser(&buf)
+	res.Header = make(http.Header)
+	res.StatusCode = 200
+
+	res.Header.Set("Content-Type", "application/json")
+	return
+}
+
+func (g graphqlGoToolsV2) headerModifier(outreq *http.Request, additionalHeaders http.Header, variableReplacer TykVariableReplacer) postprocess.HeaderModifier {
+	return func(header http.Header) {
+		for key := range additionalHeaders {
+			if header.Get(key) == "" {
+				header.Set(key, additionalHeaders.Get(key))
+			}
+		}
+
+		for key := range header {
+			val := variableReplacer(outreq, header.Get(key), false)
+			header.Set(key, val)
+		}
+	}
+}
+
+func (g graphqlGoToolsV2) returnErrorsFromUpstream(proxyOnlyCtx *GraphQLProxyOnlyContextValues, resultWriter *graphql.EngineResultWriter, seekReadCloser SeekReadCloserFunc) error {
+	body, err := seekReadCloser(proxyOnlyCtx.upstreamResponse.Body)
+	if body == nil {
+		// Response body already read by graphql-go-tools, and it's not re-readable. Quit silently.
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	responseBody, err := io.ReadAll(body)
+	if err != nil {
+		return err
+	}
+	// graphql-go-tools error message format: {"errors": [...]}
+	// Insert the upstream error into the first error message.
+	result, err := jsonparser.Set(resultWriter.Bytes(), responseBody, "errors", "[0]", "extensions")
+	if err != nil {
+		return err
+	}
+	resultWriter.Reset()
+	_, err = resultWriter.Write(result)
+	return err
+}
+
+type reverseProxyPreHandlerV2 struct {
+	ctxRetrieveGraphQLRequest ContextRetrieveRequestV2Func
+	apiDefinition             *apidef.APIDefinition
+	httpClient                *http.Client
+	newReusableBodyReadCloser NewReusableBodyReadCloserFunc
+}
+
+func (r *reverseProxyPreHandlerV2) PreHandle(params ReverseProxyParams) (reverseProxyType ReverseProxyType, err error) {
+	r.httpClient.Transport = NewGraphQLEngineTransport(
+		DetermineGraphQLEngineTransportType(r.apiDefinition),
+		params.RoundTripper,
+		r.newReusableBodyReadCloser,
+	)
+
+	switch {
+	case params.IsCORSPreflight:
+		if params.NeedsEngine {
```

Comment:
This needs to be changed, as there was a bug regarding `OPTIONS` pass-through and proxy-only.

See: https://github.com/TykTechnologies/tyk/pull/6221/files